mono/mini/mini-llvm-cpp.cpp

```diff
@@ -241,10 +241,114 @@ mono_llvm_set_preserveall_cc (LLVMValueRef func)
 	unwrap<Function>(func)->setCallingConv (CallingConv::PreserveAll);
 }
 
+// Note that in future versions of LLVM, CallInst and InvokeInst
+// share a CallBase parent class that would make the below methods
+// look much better
+
+void
+mono_llvm_set_call_preserveall_cc (LLVMValueRef wrapped_calli)
+{
+	Instruction *calli = unwrap<Instruction> (wrapped_calli);
+
+	if (isa<CallInst> (calli))
+		dyn_cast<CallInst>(calli)->setCallingConv (CallingConv::PreserveAll);
+	else
+		dyn_cast<InvokeInst>(calli)->setCallingConv (CallingConv::PreserveAll);
+}
+
 void
-mono_llvm_set_call_preserveall_cc (LLVMValueRef func)
+mono_llvm_set_call_nonnull_arg (LLVMValueRef wrapped_calli, int argNo)
 {
-	unwrap<CallInst>(func)->setCallingConv (CallingConv::PreserveAll);
+	Instruction *calli = unwrap<Instruction> (wrapped_calli);
+
+	if (isa<CallInst> (calli))
+		dyn_cast<CallInst>(calli)->addParamAttr (argNo, Attribute::NonNull);
+	else
+		dyn_cast<InvokeInst>(calli)->addParamAttr (argNo, Attribute::NonNull);
+}
+
+void
+mono_llvm_set_call_nonnull_ret (LLVMValueRef wrapped_calli)
+{
+	Instruction *calli = unwrap<Instruction> (wrapped_calli);
+
+	if (isa<CallInst> (calli))
+		dyn_cast<CallInst>(calli)->addAttribute (AttributeList::ReturnIndex, Attribute::NonNull);
+	else
+		dyn_cast<InvokeInst>(calli)->addAttribute (AttributeList::ReturnIndex, Attribute::NonNull);
+}
+
+void
+mono_llvm_set_func_nonnull_arg (LLVMValueRef func, int argNo)
+{
+	unwrap<Function>(func)->addParamAttr (argNo, Attribute::NonNull);
+}
+
+gboolean
+mono_llvm_is_nonnull (LLVMValueRef wrapped)
+{
+	// Argument to function
+	Value *val = unwrap (wrapped);
+	if (Argument *arg = dyn_cast<Argument> (val)) {
+		return arg->hasNonNullAttr ();
+	} else if (CallInst *calli = dyn_cast<CallInst> (val)) {
+		return calli->hasRetAttr (Attribute::NonNull);
+	} else if (InvokeInst *calli = dyn_cast<InvokeInst> (val)) {
+		return calli->hasRetAttr (Attribute::NonNull);
+	} else if (LoadInst *loadi = dyn_cast<LoadInst> (val)) {
+		return loadi->getMetadata("nonnull") != nullptr; // nonnull <index>
+	} else if (Instruction *inst = dyn_cast<Instruction> (val)) {
+		// If not a load or a function argument, the only case for us to
+		// consider is that it's a bitcast. If so, recurse on what was casted.
+		if (inst->getOpcode () == LLVMBitCast)
+			return mono_llvm_is_nonnull (wrap (inst->getOperand (0)));
```

Comment:
maybe a loop instead of recursion?

Reply:
Yeah, this is basically tail recursion. Can do.